src/game/game_init.c

```diff
@@ -683,34 +669,34 @@ void init_controllers(void) {
     gSramProbe = nuPiInitSram();
 #endif
 
-    // Loop over the 4 ports and link the controller structs to the appropriate
-    // status and pad. Interestingly, although there are pointers to 3 controllers,
-    // only 2 are connected here. The third seems to have been reserved for debug
-    // purposes and was never connected in the retail ROM, thus gPlayer3Controller
-    // cannot be used, despite being referenced in various code.
-    for (cont = 0, port = 0; port < 4 && cont < 2; port++) {
+    // Loop over the 4 ports and link the controller structs to the appropriate status and pad.
+    for (cont = 0, port = 0, lastUsedPort = 0; port < MAXCONTROLLERS && cont < MAX_ALLOWED_CONTROLLERS; port++) {
         // Is controller plugged in?
         if (gControllerBits & (1 << port)) {
             // The game allows you to have just 1 controller plugged
             // into any port in order to play the game. this was probably
             // so if any of the ports didn't work, you can have controllers
             // plugged into any of them and it will work.
-#if ENABLE_RUMBLE
-            gControllers[cont].port = port;
-#endif
             gControllers[cont].statusData = &gControllerStatuses[port];
-            gControllers[cont++].controllerData = &gControllerPads[port];
+            gControllers[cont].controllerData = &gControllerPads[port];
+            gControllers[cont].port = port;
+
+            lastUsedPort = port;
+
+            cont++;
         }
     }
-    if ((__osControllerTypes[1] == CONT_TYPE_GCN) && (gIsConsole)) {
-        gGamecubeControllerPort = 1;
-        gPlayer1Controller = &gControllers[1];
-    } else {
-        if (__osControllerTypes[0] == CONT_TYPE_GCN) {
-            gGamecubeControllerPort = 0;
-        }
-        gPlayer1Controller = &gControllers[0];
+
+    // Some flashcarts (eg. ED64p) don't let you start a ROM with a GameCube controller in port 1,
+    // so if port 1 is an N64 controller and port 2 is a GC controller, swap them.
+    if (gIsConsole && __osControllerTypes[0] == CONT_TYPE_N64 && __osControllerTypes[1] == CONT_TYPE_GCN) {
+        struct Controller temp = gControllers[0];
+        gControllers[0] = gControllers[1];
+        gControllers[1] = temp;
     }
+
+    // Disable the ports after the last used one.
+    osContSetCh(lastUsedPort + 1);
```

Comment:
this should also have `MIN` with `MAX_ALLOWED_CONTROLLERS` right? it's a weird situation but clamping to usable controllers makes sense if somebody has controllers plugged into the first and last port for no reason. we shouldn't open up port 3 or 4 unless max controllers exceeds it

Reply:
That's already covered by the loop in that function, with `cont < MAX_ALLOWED_CONTROLLERS`. Having it in `osContSetCh` with a `MIN` caused issues because it uses port numbers and `MAX_ALLOWED_CONTROLLERS ` is number of controllers. For example, having one controller plugged into port 3 or 4 when `MAX_ALLOWED_CONTROLLERS` is 2 wouldn't work.

Reply:
I don't think that's a good idea considering any port higher than 2 would cause a good degree of lag. I think it's better in that case to say that the "controller isn't connected into port 1", and then later we could consider adding a hot swap feature if the controller isnt plugged into port 1

I just think it's not necessary to allow any other ports beyond the max for super mild convenience if it has such a big drawback

Reply:
My understanding was this would only happen if you had controllers plugged into those ports, which on one hand having something with a few hundred μs extra lag is better than having something completely nonfunctional. That said, people that have controllers plugged into those ports simply out of laziness or cluelessness will experience additional lag for no good reason.

Reply:
Being able to use any port for player 1 is vanilla behavior, I don't think we should remove that entirely. Maybe that can be a separate config define?

Reply:
any port for player 1 is vanilla? I did not realize that actually. but yeah after the first controller is connected I think that the number of supported players should dictate the maximum port. like if you are plugged into 1 and 4, just poll 1 if you only support single player

Reply:
I agree wtih cozies

Reply:
That's what the define would do if you set it to 1, but it needs to be set to 2 for the ED64p thing.

Should the ED64p check bypass the define and ignore the N64 controller completely and only use the first GameCube controller as Player 1 instead of swapping them, if the define is set to 1?